File: tests/oracle/blockstruct_dowhile_absorb_1204.cc
/*
 * BLOCKACTION-DOWHILE-ABSORB-0001: locked Ghidra 12.0.4
 * CollapseStructure::collapseAll (blockaction.cc:1877-1893) oracle,
 * covering the do-while absorption chain at the newBlockList force step:
 *
 *   BlockGraph::newBlockList (block.cc:1758-1774) captures
 *   outforce = nodes.back()->sizeOut() (and out0 when binary) BEFORE
 *   identifyInternal, then forceOutputNum(outforce) (block.cc:880-889)
 *   resurrects a chain-internal back-edge as a composite SELF edge labeled
 *   f_loop_edge|f_back_edge, and forceFalseEdge (block.cc:1204-1217)
 *   preserves which branch was the false path (out0 internal → self).
 *   That self edge is what ruleBlockDoWhile (cc:1555) absorbs — the
 *   main @321a/@3225/@28ec/@28f7 latch-pair shape: ruleBlockGoto
 *   (newBlockIfGoto) → ruleBlockCat (force self edge) → ruleBlockDoWhile.
 *
 * Cases:
 *   1. latch_pair_goto_dowhile — the main @321a topology (head exits to
 *      body+latch-exit target, latch exits back to head and to the shared
 *      target): TraceDAG/selectGoto marks the head's exit, ruleBlockGoto
 *      wraps an if-goto, cat merges it with the latch, forceOutputNum
 *      restores the back-edge as the composite self edge, ruleBlockDoWhile
 *      absorbs the latch. Final tree must contain dowhile(list{if(goto),latch}).
 *   2. chain_tail_backedge_dowhile — no goto at all: a straight chain whose
 *      tail branches back into the chain. cat merges the chain,
 *      forceOutputNum adds the self edge, dowhile absorbs. This isolates
 *      the pure newBlockList force semantic.
 *   3. tail_out0_internal_flip — the tail's out(0) points back INTO the
 *      chain (out0 is internalized): forceFalseEdge's parent==this arm
 *      must swap the self edge to slot 0, and ruleBlockDoWhile must
 *      negate (cc:1566-1569) — observable as the flip flag.
 */

#include <bits/stdc++.h>

// Test-only access (same scheme as blockstruct_goto_cascade_1204.cc):
// BlockGraph's helper members (structureLoops, getList, addBlock) are
// implicitly private; `class -> struct` + `private -> public` opens them
// in this TU only.
#define private public
#define class struct
#include "architecture.hh"
#include "block.hh"
#include "blockaction.hh"
#include "database.hh"
#include "funcdata.hh"
#include "libdecomp.hh"
#include "marshal.hh"
#include "op.hh"
#include "translate.hh"
#include "type.hh"
#include "typeop.hh"
#include "varnode.hh"
#undef class
#undef private

using namespace ghidra;

static std::string flagsText(uint4 label)
{
  std::string out;
  if (label & FlowBlock::f_goto_edge) out += 'g';
  if (label & FlowBlock::f_loop_edge) out += 'l';
  if (label & FlowBlock::f_defaultswitch_edge) out += 'd';
  if (label & FlowBlock::f_irreducible) out += 'i';
  if (label & FlowBlock::f_tree_edge) out += 't';
  if (label & FlowBlock::f_forward_edge) out += 'f';
  if (label & FlowBlock::f_cross_edge) out += 'c';
  if (label & FlowBlock::f_back_edge) out += 'b';
  if (label & FlowBlock::f_loop_exit_edge) out += 'x';
  if (out.empty()) out.push_back('-');
  return out;
}

// Synthetic vertex: plain FlowBlock topology, isComplex()==false so the
// condition-bearing blocks fold exactly like real basic blocks.
struct Vertex : public FlowBlock {
  int id;
  explicit Vertex(int id_) : FlowBlock(), id(id_) {}
  virtual block_type getType(void) const { return t_basic; }
  virtual bool isComplex(void) const { return false; }
  virtual void printHeader(ostream &s) const { s << 'b' << id; }
};

class Graph {
public:
  BlockGraph graph;
  std::vector<FlowBlock *> created;

  FlowBlock *makeBlock(void)
  {
    FlowBlock *bl = new Vertex((int4)created.size());
    graph.addBlock(bl);
    created.push_back(bl);
    return bl;
  }

  void edge(FlowBlock *from, FlowBlock *to)
  {
    graph.addEdge(from, to);
  }

  std::string nameOf(FlowBlock *bl, const std::vector<FlowBlock *> &toplist) const
  {
    for (int4 i = 0; i < created.size(); ++i)
      if (created[i] == bl) return string("b") + char('0' + i);
    for (int4 i = 0; i < toplist.size(); ++i)
      if (toplist[i] == bl) return string("c") + char('0' + i);
    return "b?";
  }

  void run(const std::string &caseName)
  {
    // ActionBlockStructure entry state (cc:2177-2180): the structure copy
    // inherits spanning-tree/back-edge labels; reproduce by running
    // structureLoops on the same graph.
    std::vector<FlowBlock *> rootlist;
    graph.structureLoops(rootlist);
    graph.clearVisitCount();

    CollapseStructure collapse(graph);
    collapse.collapseAll();

    // ---- observation ----
    std::vector<FlowBlock *> toplist = graph.getList();
    std::cout << "case " << caseName << '\n';
    std::cout << "list=";
    for (int4 i = 0; i < (int4)toplist.size(); ++i) {
      if (i != 0) std::cout << ',';
      std::cout << 'c' << i << '(' << FlowBlock::typeToName(toplist[i]->getType()) << ')';
    }
    std::cout << '\n';
    for (int4 i = 0; i < (int4)toplist.size(); ++i) {
      FlowBlock *bl = toplist[i];
      std::cout << 'c' << i << " type=" << FlowBlock::typeToName(bl->getType())
                << " gotoout=" << (bl->hasInteriorGoto() ? 1 : 0)
                << " out=[";
      for (int4 j = 0; j < bl->sizeOut(); ++j) {
        if (j != 0) std::cout << ' ';
        std::cout << j << ':' << flagsText(bl->outofthis[j].label)
                  << "->" << nameOf(bl->getOut(j), toplist);
      }
      std::cout << "]\n";
      std::cout << "tree c" << i << ":\n";
      dumpTree(bl, toplist, 1);
    }
    for (int4 i = 0; i < (int4)created.size(); ++i) {
      FlowBlock *bl = created[i];
      std::cout << 'b' << i << " out=[";
      for (int4 j = 0; j < bl->sizeOut(); ++j) {
        if (j != 0) std::cout << ' ';
        std::cout << j << ':' << flagsText(bl->outofthis[j].label);
      }
      std::cout << "] flip=" << (bl->getFlipPath() ? 1 : 0)
                << " listed=" << (std::find(toplist.begin(), toplist.end(), bl) != toplist.end() ? 1 : 0)
                << '\n';
    }
    std::cout << "end\n";
  }

  // Recursive structure-tree projection (same scheme as
  // blockstruct_goto_cascade_1204.cc): child order is the factory node
  // order. t_goto children are NOT dumped (documented normalization).
  void dumpTree(FlowBlock *bl, const std::vector<FlowBlock *> &toplist, int4 depth) const
  {
    for (int4 i = 0; i < depth; ++i) std::cout << "  ";
    std::cout << FlowBlock::typeToName(bl->getType());
    if (bl->getType() == FlowBlock::t_if) {
      BlockIf *bif = (BlockIf *)bl;
      if (bif->getGotoTarget() != (FlowBlock *)0)
        std::cout << " gototarget=" << nameOf(bif->getGotoTarget(), toplist);
    }
    if (bl->hasInteriorGoto()) std::cout << " gotoout=1";
    if (bl->getFlipPath()) std::cout << " flip=1";
    std::cout << '\n';
    if (bl->getType() == FlowBlock::t_goto) return;
    BlockGraph *g = dynamic_cast<BlockGraph *>(bl);
    if (g == (BlockGraph *)0) return;
    for (int4 i = 0; i < g->getSize(); ++i)
      dumpTree(g->getBlock(i), toplist, depth + 1);
  }
};

int main(int argc, char **argv)
{
  std::string filter = (argc > 1) ? argv[1] : "";

  if (filter.empty() || filter == "latch_pair_goto_dowhile") {
    Graph g;
    FlowBlock *b0 = g.makeBlock();
    FlowBlock *b1 = g.makeBlock();
    FlowBlock *b2 = g.makeBlock();
    FlowBlock *b3 = g.makeBlock();
    g.edge(b0, b1);
    g.edge(b1, b2);
    g.edge(b1, b3);
    g.edge(b2, b3);
    g.edge(b2, b1);
    g.run("latch_pair_goto_dowhile");
  }
  if (filter.empty() || filter == "chain_tail_backedge_dowhile") {
    Graph g;
    FlowBlock *b0 = g.makeBlock();
    FlowBlock *b1 = g.makeBlock();
    FlowBlock *b2 = g.makeBlock();
    FlowBlock *b3 = g.makeBlock();
    g.edge(b0, b1);
    g.edge(b1, b2);
    g.edge(b2, b3);
    g.edge(b2, b1);
    g.run("chain_tail_backedge_dowhile");
  }
  if (filter.empty() || filter == "tail_out0_internal_flip") {
    Graph g;
    FlowBlock *b0 = g.makeBlock();
    FlowBlock *b1 = g.makeBlock();
    FlowBlock *b2 = g.makeBlock();
    FlowBlock *b3 = g.makeBlock();
    FlowBlock *b9 = g.makeBlock();
    g.edge(b0, b1);
    g.edge(b0, b9);
    g.edge(b1, b2);
    g.edge(b2, b1);
    g.edge(b2, b3);
    g.run("tail_out0_internal_flip");
  }
  return 0;
}

File: tests/oracle/blockstruct_dowhile_absorb_1204.rs
// BLOCKACTION-DOWHILE-ABSORB-0001: Rugra comparand for the locked Ghidra
// 12.0.4 CollapseStructure::collapseAll oracle (blockaction.cc:1877-1893).
// Mirrors tests/oracle/blockstruct_dowhile_absorb_1204.cc case for case:
// the same synthetic graphs are built through the production BlockGraph
// APIs (add_block/add_edge), CollapseStructure::collapse_all runs (the
// 5-step faithful path), and the post-run state prints in the shared
// observation format (top-level list, structure tree, per-vertex
// surviving out-edge labels / flip flag / listed flag).
//
// The covered semantic is the newBlockList force step (block.cc:1758-1774):
// forceOutputNum(outforce) resurrects a chain-internal back-edge as a
// composite SELF edge (f_loop_edge|f_back_edge, block.cc:880-889) so
// ruleBlockDoWhile (cc:1555) can absorb the latch, and forceFalseEdge
// (block.cc:1204-1217) preserves the false branch (out0 internal -> self).

use std::sync::{Arc, RwLock};

use rugra::address::Address;
use rugra::block::{
    edge_flags as ef, block_flags as bf, BlockBasic, BlockCondition, BlockDoWhile, BlockGraph,
    BlockIf, BlockInfLoop, BlockList, BlockType, BlockWhileDo, FlowBlock,
};

type BlockRef = Arc<RwLock<dyn FlowBlock + Send + Sync>>;

fn flags_text(label: u32) -> String {
    let mut out = String::new();
    if label & ef::F_GOTO_EDGE != 0 { out.push('g'); }
    if label & ef::F_LOOP_EDGE != 0 { out.push('l'); }
    if label & ef::F_DEFAULTSWITCH_EDGE != 0 && label & ef::F_TREE_EDGE == 0 { out.push('d'); }
    if label & ef::F_IRREDUCIBLE_EDGE != 0 { out.push('i'); }
    if label & ef::F_TREE_EDGE != 0 { out.push('t'); }
    if label & ef::F_FORWARD_EDGE != 0 { out.push('f'); }
    if label & ef::F_CROSS_EDGE != 0 { out.push('c'); }
    if label & ef::F_BACK_EDGE != 0 { out.push('b'); }
    if label & ef::F_LOOP_EXIT_EDGE != 0 { out.push('x'); }
    if out.is_empty() { out.push('-'); }
    out
}

// FlowBlock::typeToName (block.cc:671-703) — note t_if prints "properif".
fn type_name(bt: BlockType) -> &'static str {
    match bt {
        BlockType::Plain => "plain",
        BlockType::Basic => "basic",
        BlockType::Graph => "graph",
        BlockType::Copy => "copy",
        BlockType::Goto => "goto",
        BlockType::MultiGoto => "multigoto",
        BlockType::List => "list",
        BlockType::Condition => "condition",
        BlockType::If => "properif",
        BlockType::WhileDo => "whiledo",
        BlockType::DoWhile => "dowhile",
        BlockType::Switch => "switch",
        BlockType::InfLoop => "infloop",
    }
}

struct Graph {
    graph: BlockGraph,
    created: Vec<BlockRef>,
    base: u64,
}

impl Graph {
    fn new(base: u64) -> Self {
        Graph { graph: BlockGraph::new(), created: Vec::new(), base }
    }

    fn make_block(&mut self) -> BlockRef {
        // Unique index AND address per block: Rugra's collapse model keys
        // block identity on FlowBlock::get_index() (Ghidra uses parent
        // pointers), and loop discovery keys on addresses — all-zero
        // indices/addresses make every self-loop test vacuously true.
        let idx = self.created.len() as i32;
        let bl: BlockRef = Arc::new(RwLock::new(BlockBasic::new(
            idx,
            Address::new(self.base + (idx as u64) * 0x10),
        )));
        self.graph.add_block(bl.clone());
        self.created.push(bl.clone());
        bl
    }

    fn edge(&mut self, from: &BlockRef, to: &BlockRef) {
        self.graph.add_edge(from.clone(), to.clone());
    }

    fn name_of(&self, bl: &BlockRef, toplist: &[BlockRef]) -> String {
        for (i, cand) in self.created.iter().enumerate() {
            if Arc::ptr_eq(cand, bl) {
                return format!("b{}", i);
            }
        }
        for (i, cand) in toplist.iter().enumerate() {
            if Arc::ptr_eq(cand, bl) {
                return format!("c{}", i);
            }
        }
        "b?".to_string()
    }

    fn run(&mut self, case_name: &str) {
        // ActionBlockStructure entry state (cc:2177-2180): the structure
        // copy inherits spanning-tree/back-edge labels; reproduce by
        // running structure_loops on the graph first, exactly as the C++
        // fixture's run() does via structureLoops + clearVisitCount.
        let mut rootlist: Vec<BlockRef> = Vec::new();
        let _ = self.graph.structure_loops(&mut rootlist);
        // clearVisitCount (cc:2181) is folded into order_loop_bodies on the
        // Rugra side (see collapse_all_5step cc:1879-1884 note).

        // CollapseStructure::collapseAll via the public 5-step driver.
        let mut collapse = rugra::blockaction::CollapseStructure::new(
            &mut self.graph,
            case_name,
        );
        collapse.collapse_all();

        // ---- observation (mirrors the C++ fixture byte for byte) ----
        let toplist: Vec<BlockRef> = self.graph.blocks.clone();
        println!("case {}", case_name);
        print!("list=");
        for (i, bl) in toplist.iter().enumerate() {
            if i != 0 { print!(","); }
            let bt = bl.read().unwrap().get_type();
            print!("c{}({})", i, type_name(bt));
        }
        println!();
        for (i, bl) in toplist.iter().enumerate() {
            let r = bl.read().unwrap();
            print!(
                "c{} type={} gotoout={} out=[",
                i,
                type_name(r.get_type()),
                if r.get_flags() & bf::INTERIOR_GOTOOUT != 0 { 1 } else { 0 }
            );
            for j in 0..r.size_out() {
                if j != 0 { print!(" "); }
                let label = r.get_out(j).map(|e| e.flags).unwrap_or(0);
                let target = r.get_out(j).map(|e| e.point.clone());
                let tname = match target {
                    Some(t) => self.name_of(&t, &toplist),
                    None => "b?".to_string(),
                };
                print!("{}:{}->{}", j, flags_text(label), tname);
            }
            println!("]");
            drop(r);
            println!("tree c{}:", i);
            self.dump_tree(bl, &toplist, 1);
        }
        for (i, bl) in self.created.iter().enumerate() {
            let r = bl.read().unwrap();
            print!("b{} out=[", i);
            for j in 0..r.size_out() {
                if j != 0 { print!(" "); }
                let label = r.get_out(j).map(|e| e.flags).unwrap_or(0);
                print!("{}:{}", j, flags_text(label));
            }
            println!(
                "] flip={} listed={}",
                if r.get_flags() & bf::FLIP_PATH != 0 { 1 } else { 0 },
                if toplist.iter().any(|t| Arc::ptr_eq(t, bl)) { 1 } else { 0 }
            );
        }
        println!("end");
    }

    // Recursive structure-tree projection mirroring the C++ dumpTree: child
    // order is the factory node order. BlockGoto children are NOT dumped
    // (Rugra's newBlockGoto wraps in place — documented normalization on
    // both sides).
    fn dump_tree(&self, bl: &BlockRef, toplist: &[BlockRef], depth: usize) {
        let r = bl.read().unwrap();
        for _ in 0..depth { print!("  "); }
        print!("{}", type_name(r.get_type()));
        let (bt, flags) = (r.get_type(), r.get_flags());
        if bt == BlockType::If {
            if let Some(bi) = r.as_any().downcast_ref::<BlockIf>() {
                if let Some(ref gt) = bi.goto_target {
                    print!(" gototarget={}", self.name_of(gt, toplist));
                }
            }
        }
        if flags & bf::INTERIOR_GOTOOUT != 0 { print!(" gotoout=1"); }
        if flags & bf::FLIP_PATH != 0 { print!(" flip=1"); }
        println!();
        if bt == BlockType::Goto {
            return;
        }
        let children: Vec<BlockRef> = match bt {
            BlockType::List => r
                .as_any()
                .downcast_ref::<BlockList>()
                .map(|b| b.children.clone())
                .unwrap_or_default(),
            BlockType::If => {
                let mut v = Vec::new();
                if let Some(bi) = r.as_any().downcast_ref::<BlockIf>() {
                    v.push(bi.condition.clone());
                    // IfGoto normalization: newBlockIfGoto (block.cc:1799-
                    // 1816) installs ONLY the condition in the composite
                    // (nodes={cond}); the body stays external as the single
                    // structured out-edge. Rugra's BlockIf keeps a
                    // non-Option if_body placeholder (= the condition
                    // itself) for the if-goto wrap — do not dump it (same
                    // spirit as the BlockGoto children normalization).
                    if bi.goto_target.is_none() {
                        v.push(bi.if_body.clone());
                        if let Some(ref eb) = bi.else_body {
                            v.push(eb.clone());
                        }
                    }
                }
                v
            }
            BlockType::WhileDo => r
                .as_any()
                .downcast_ref::<BlockWhileDo>()
                .map(|b| vec![b.condition.clone(), b.body.clone()])
                .unwrap_or_default(),
            BlockType::DoWhile => r
                .as_any()
                .downcast_ref::<BlockDoWhile>()
                .map(|b| vec![b.condition.clone()])
                .unwrap_or_default(),
            BlockType::Condition => r
                .as_any()
                .downcast_ref::<BlockCondition>()
                .map(|b| vec![b.first.clone(), b.second.clone()])
                .unwrap_or_default(),
            BlockType::InfLoop => r
                .as_any()
                .downcast_ref::<BlockInfLoop>()
                .map(|b| vec![b.body.clone()])
                .unwrap_or_default(),
            _ => Vec::new(),
        };
        drop(r);
        for child in children {
            self.dump_tree(&child, toplist, depth + 1);
        }
    }
}

fn main() {
    // Optional argv[1] filter: run a single named case (debug convenience).
    let filter = std::env::args().nth(1);
    macro_rules! case {
        ($name:expr, $body:block) => {
            if filter.is_none() || filter.as_deref() == Some($name) $body
        };
    }
    case!("latch_pair_goto_dowhile", {
        // 1. latch_pair_goto_dowhile — the main @321a topology: head exits
        // to body + shared target, latch exits to target + back to head.
        let mut g = Graph::new(0x1000);
        let b0 = g.make_block();
        let b1 = g.make_block();
        let b2 = g.make_block();
        let b3 = g.make_block();
        g.edge(&b0, &b1);
        g.edge(&b1, &b2);
        g.edge(&b1, &b3);
        g.edge(&b2, &b3);
        g.edge(&b2, &b1);
        g.run("latch_pair_goto_dowhile");
    });
    case!("chain_tail_backedge_dowhile", {
        // 2. chain_tail_backedge_dowhile — no goto: straight chain whose
        // tail branches back into the chain; cat + forceOutputNum + dowhile.
        let mut g = Graph::new(0x2000);
        let b0 = g.make_block();
        let b1 = g.make_block();
        let b2 = g.make_block();
        let b3 = g.make_block();
        g.edge(&b0, &b1);
        g.edge(&b1, &b2);
        g.edge(&b2, &b3);
        g.edge(&b2, &b1);
        g.run("chain_tail_backedge_dowhile");
    });
    case!("tail_out0_internal_flip", {
        // 3. tail_out0_internal_flip — tail's out(0) points back INTO the
        // chain: forceFalseEdge's parent==this arm + slot-0 dowhile negate.
        let mut g = Graph::new(0x3000);
        let b0 = g.make_block();
        let b1 = g.make_block();
        let b2 = g.make_block();
        let b3 = g.make_block();
        let b9 = g.make_block();
        g.edge(&b0, &b1);
        g.edge(&b0, &b9);
        g.edge(&b1, &b2);
        g.edge(&b2, &b1);
        g.edge(&b2, &b3);
        g.run("tail_out0_internal_flip");
    });
}
